cpp/src/trajectory/compute_bounding_boxes.cu

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/table/table.hpp>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cuspatial/detail/trajectory.hpp>
+
+namespace cuspatial {
+namespace experimental {
+
+namespace {
+
+template <typename Element>
+__global__ void compute_bounding_boxes_kernel(
```

Comment:
This kernel finds the min/max x/y for each route in a trajectory (which is a group of routes). Assuming it's possible for cub to use thrust tuples/zip_iterators, we could do this with `cub::DeviceSegmentedReduce` using the offsets computed by `derive_trajectories`.

Alternatively we can change the API to accept the original `object_id` column instead of the offsets from `derive_trajectories`, and use `thrust::reduce_by_key`, which is equivalent to this in cudf:

```python
points_df = cudf.DataFrame({
  'id': [0, 0, 1, 1],
  'x':  [1, 2, 3, 0],
  'y':  [2, 3, 4, 1] })

trajectory_bboxes = points_df \
     .groupby(by='id', as_index=False, level=1) \
     .agg({ 'id': 'min', 'x': ['min', 'max'], 'y': ['min', 'max'] })

print(trajectory_bboxes)

#   id   x       y    
#      min max min max
# 0  0   1   2   2   3
# 1  1   0   3   1   4

```

Reply:
Hmmm, reduce_by_key is the most intuitive. If it allows us to do away with computing offsets altogether, I'm all for that.

Reply:
The `thrust::reduce_by_key` re-implementations are definitely cleaner. Removing the offsets means we have to call thrust `unique_by_key()` to determine the size of the output columns. This seems like an acceptable cost if we're trying to save the memory allocated for the offsets.

On the other hand if we'd rather not repeat work, we could precompute the offsets. Doesn't thrust's reduce_by_key have to figure out offsets from the keys internally anyway? If so, we might be able to skip that too and call `cub::DeviceSegmentedReduce` directly.